Address global HPKE config - Issue #11 (#21)
wangshan/draft-wang-ppm-dap-taskprov: draft-wang-ppm-dap-taskprov.md

The behavior of each protocol participant is determined by whether or not they
opt in to a task.

## Supporting a Global HPKE Configuration {#global-hpke-config}

> 

Reviewer: editorial: I suggest reworking this so that all of the motivation is in the first paagragh, then all normative text follows:

> Clients need to know the HPKE configuration prior to uploading reports, but 
>
> To mitigate this issue, Aggregators SHOULD ...
>
> In addition, Clients SHOULD NOT ...

Author: Good point, there are three paragraphs: 1. describe the issue of specifying `task_id` in `/hpke_config`, 2. Aggregator's recommended behavior, 3. Client's recommended behavior.

In DAP, Clients need to know the HPKE configuration of each Aggregator before

> sending reports (See HPKE Configuration Request in

Reviewer: ```suggestion
sending reports. (See HPKE Configuration Request in
```

{{!DAP=I-D.draft-ietf-ppm-dap-02}}.)

If `task_prov` extension is supported in a DAP deployment, each Aggregator

> SHOULD support a global HPKE configuration independent of Task ID, because

Reviewer: ```suggestion
SHOULD support a global HPKE configuration independent of the task ID, because
```

a task is only configured in an Aggregator after the Aggregator has received
the first Client report that contains the `task_prov` extension.

> If a Client wants to query an Aggregator's `[aggregator]/hpke_config` endpoint,

Reviewer: There is no reason to repeat the endpoint format here. The more we can refer to the DAP draft as the source of truth, the better.

The behavior of each protocol participant is determined by whether or not they
opt in to a task.

## Supporting HPKE Configurations Independent of Tasks {#hpke-config-no-task-id}

In DAP, Clients need to know the HPKE configuration of each Aggregator before
sending reports. (See HPKE Configuration Request in
{{!DAP=I-D.draft-ietf-ppm-dap-02}}.) However, in a DAP deployment that supports

> `task_prov` extension, if a Client requests the Aggregator's HPKE configuration

Reviewer: ```suggestion
the `task_prov` extension, if a Client requests the Aggregator's HPKE configuration
```

with the task ID computed as described in {{construct-task-id}}, the
task ID may not be configured in the Aggregator yet, because the Aggregator is
still waiting for the first Client report with the `task_prov` extension
to arrive.

To mitigate this issue, if an Aggregator wants to support `task_prov` extension,
it SHOULD choose which HPKE configuration to advertise to Clients independent of
the task ID. It MAY continue to support per-task HPKE configurations for other
tasks that are configured out-of-band.

> In addition, if a Client wants to include `task_prov` extension in its report,

Reviewer: ```suggestion
In addition, if a Client wants to include the `task_prov` extension in its report,
```

The behavior of each protocol participant is determined by whether or not they
opt in to a task.

## Supporting a Global HPKE Configuration {#global-hpke-config}

In DAP, Clients need to know the HPKE configuration of each Aggregator before
sending reports. (See HPKE Configuration Request in
{{!DAP=I-D.draft-ietf-ppm-dap-02}}.) However, in a DAP deployment that supports

> `task_prov` extension, if a Client queries an Aggregator's `/hpke_config`
> endpoint with a task ID computed as described in {{construct-task-id}}, the

Reviewer: It's best to avoid writing text that refers to specific constructions in the base protocol. Suppose for example that the core protocol changed the endpoint to `/get_that_cool_hpke_config`: Then we would have to update this reference in-kind. My suggestion avoids this.
```suggestion
`task_prov` extension, if a Client requests the Aggregator's HPKE configuration
with the task ID computed as described in {{construct-task-id}}, the
```

task ID may not be configured in the Aggregator yet, because the Aggregator is
still waiting for the first Client report with the `task_prov` extension
to arrive.

To mitigate this issue, if an Aggregator wants to support `task_prov` extension,

> it SHOULD support a global HPKE configuration independent of the task ID, for all

Reviewer: Re: https://github.com/wangshan/draft-wang-ppm-dap-taskprov/pull/21#issuecomment-1271070699

The problem is the guidance "SHOULD support a global HPKE configuration". We don't want to limit the Aggregator to just one configuration, for the following reasons:
1. It may want to support multiple HPKE cipher suites (say, X25519 and P256)
2. It may want to support multiple HPKE keys at the same time (say, to account for clients that may have used an older config)

Something like this might be better: "SHOULD choose which HPKE configuration to advertise to clients independent of the task ID."

Author: I see, I didn't realize that is not clear. Maybe we should avoid the phrase "a global HPKE configuration".

I applied your suggestion, and also updated the title of this section to `Supporting HPKE Configurations Independent of Tasks {#hpke-config-no-task-id}`. Let me know if that doesn't make sense.

tasks that are configured with `task_prov` extension. It MAY continue to support
per-task HPKE configurations for other tasks that are configured out-of-band.

In addition, if a Client wants to include `task_prov` extension in its report,
it SHOULD NOT specify the `task_id` parameter when querying an Aggregator's

> `/hpke_config` endpoint, because the task ID may not have been configured
> in the Aggregator yet.

Reviewer: I think the reason is clear at this point.

```suggestion
`/hpke_config` endpoint.
```

Author: Also changed `/hpke_config` to "when requesting the HPKE configuration from an Aggregator".
